### PR TITLE
Terminate the local shell when a session closes 

### LINKED ISSUE #4469: Child processes should be killed when Teleport session ends
### Feature Request
Teleport should act like SSH, and kill any child processes when the connection is closed (intentionally or not).

### Motivation
Processes are not completely cleaned up when a Teleport session ends. Most cases are when the connection ends abruptly, such as a network drop, but customer did mentioned they did not confirm that that is always the case.

The majority of these cases are when the engineers are running 'rails console' (effectively an interactive ruby interpreter that's loaded all the libraries and dependencies of a project), because it consumes a lot of memory, and there may be multiple old consoles running on a machine.

* Unable to 'kill <pid>' the process, and have to do 'kill -9 <pid>' to get it to terminate.
* 'service teleport reload' will successfully kill the child processes (once all teleport connections have ended).
* The process's parent has been reassigned to pid1
* strace reveals that the main process is FUTEX_WAITing for another thread to return. strace on that thread reveals that it's waiting from input from stdin.

(Internal reference, more details in Zendesk ticket: [#1951](https://gravitational.zendesk.com/agent/tickets/1951))

### Who's it for?
OSS User, Pro, Enterprise

### LINKED ISSUE #6245: Some processes in some setup not killed after disconnect
### Description

**What happened**:
after running some commands while connected via tsh and closing terminal tab or browser tab some processes persist forever

**What you expected to happen**:
These processes should be killed like in using regular ssh

### Reproduction Steps

starting using ssh: ssh -> "sudo su -" -> vi qwe 

ps auxf
```
root        1353  0.0  0.8  13664  8780 ?        Ss   14:32   0:00  \_ sshd: ubuntu [priv]
ubuntu      1463  0.0  0.5  13672  5100 ?        S    14:32   0:00      \_ sshd: ubuntu@pts/2
ubuntu      1464  0.0  0.5  10540  5496 pts/2    Ss   14:32   0:00          \_ -bash
root        1962  0.0  0.4  11168  4816 pts/2    S+   14:44   0:00              \_ sudo su -
root        1963  0.0  0.0  11168   748 pts/0    Ss   14:44   0:00                  \_ sudo su -
root        1964  0.0  0.4  10128  4264 pts/0    S    14:44   0:00                      \_ su -
root        1965  0.1  0.5  10028  5060 pts/0    S    14:44   0:00                          \_ -bash
root        1976  0.2  0.9  23836  9908 pts/0    S+   14:44   0:00                              \_ vi qwe
```

after closing termianl tab with ssh, all processes from this tree are killed

now using tsh: tsh ssh ubuntu@ip-172-30-48-21 -> "sudo su -" -> vi asd 
```
root        1992  0.2  4.9 1070696 49380 ?       Sl   14:45   0:00  \_ /usr/local/bin/teleport exec
ubuntu      2077  0.2  0.5  10036  5076 pts/0    Ss   14:45   0:00      \_ -bash
root        2086  0.0  0.4  11160  4684 pts/0    S+   14:45   0:00          \_ sudo su -
root        2087  0.0  0.0  11160   744 pts/2    Ss   14:45   0:00              \_ sudo su -
root        2089  0.0  0.4  10128  4240 pts/2    S    14:45   0:00                  \_ su -
root        2090  0.1  0.4  10028  4980 pts/2    S    14:45   0:00                      \_ -bash
root        2101  0.0  0.9  23836  9888 pts/2    S+   14:45   0:00                          \_ vi asd
```
after closing terminal tab with tsh

after a minute ps auxf:
```
root        2086  0.0  0.4  11160  4684 ?        S    14:45   0:00 sudo su -
root        2087  0.0  0.0  11160   744 pts/2    Ss   14:45   0:00  \_ sudo su -
root        2089  0.0  0.4  10128  4240 pts/2    S    14:45   0:00      \_ su -
root        2090  0.0  0.4  10028  4980 pts/2    S    14:45   0:00          \_ -bash
root        2101  0.0  0.9  23836  9888 pts/2    S+   14:45   0:00              \_ vi asd
```
and this processes are never killed

### Server Details
teleport.yaml

```
teleport:
  log:
    output: stdout
    severity: INFO

  auth_token: REDACTED

  auth_servers:
    - REDACTED:3025

  storage:
    type: dir

auth_service:
  enabled: no


ssh_service:
  enabled: yes
  pam:
    enabled: yes

  listen_addr: 0.0.0.0:3022

  labels:
    cluster: box
    owner: aiwaniuk

  permit_user_env: false

proxy_service:
  enabled: no
```

fresh ubuntu 20.04 from AWS
```
ubuntu@ip-172-30-48-21:~$ uname -a
Linux ip-172-30-48-21 5.4.0-1038-aws #40-Ubuntu SMP Fri Feb 5 23:50:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
ubuntu@ip-172-30-48-21:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:  Ubuntu 20.04.2 LTS
Release:  20.04
Codename: focal
```
one modification:
```
/etc/sudoers
"Defaults        env_reset" -> "Defaults        env_reset, use_pty"
```

both proxy and auth and node are running 6.0.02
```
ubuntu@ip-172-30-48-21:~$ teleport version
Teleport Enterprise v6.0.2 git: go1.15.5b5
```

### LINKED ISSUE #13335: TestAgentForward leaves orphaned shell processes behind
While working on fixing the flakiness in `TestAgentForward` (#13166) I noticed that it leaves behind a shell process each time it is run.

To reproduce, I pull up Activity Monitor on mac, sort by process name in descending order, and scroll to the top.

Each time I run the test with the following command, I get a new zsh process that stays running even after the test completes:

```
$ go test -run 'TestAgentForward$' -race -count=1 ./lib/srv/regular 
```

### PR DESCRIPTION
**Pull request by jakule**
When a client sends an SSH session before the shell process ends, Teleport sends the SIGKILL signal to the subprocess. This causes the still-leaving shell to be resigned to PID 1 and live forever.
This PR tried to stop the terminal first before killing the subprocess to prevent the terminal from living forever.

Closes https://github.com/gravitational/teleport/issues/13335
Closes https://github.com/gravitational/teleport/issues/6245
Closes https://github.com/gravitational/teleport/issues/4469

### CODE REVIEW

File: lib/srv/reexec.go [hunk 1]
```diff
@@ -418,6 +422,37 @@ func RunCommand() (errw io.Writer, code int, err error) {
...
+		// message here because the shell has successfully started. If an error
+		// occurred during shell execution or the shell exits with an error (like
+		// running exit 2), the shell will print an error if appropriate and return
+		// an exit code.
+		err := cmd.Wait()
+
+		terminateChan <- err
+	}()
+
+	err := <-terminateChan
```
Comment (zmb3) [inline lib/srv/reexec.go:452]: You're doing 2 writes to `terminateChan` and only 1 read. Doesn't that mean that one of these goroutines will leak, and that you may miss an error depending on timing?

Comment (jakule) [reply lib/srv/reexec.go:452]: > Doesn't that mean that one of these goroutines will leak

Yes, but after one goroutine finishes, the process ends, so leaking anything at this point is not a huge deal.

> and that you may miss an error depending on timing

This is more complicated. The ideal scenario happens when the shell process end, and we don't need to send kill signals to it. The scenario that this PR tries to fix happens when the SSH session sends on the client side, and someone needs to kill the "abandoned shell". If we try to kill the shell after the shell process ends, we don't care about any returned errors. I also need to check what is logged in our audit log when that happens (and TBH, I'm not sure what the correct value should be). If the shell process ends without an error, then we don't want to override that with the potential error returned from the other goroutine. I'll add a comment to explain that, as this is probably "not self-explanatory" just by looking at the code. I'm also not saying that the current version is 100% working.

Comment (zmb3) [reply lib/srv/reexec.go:452]: Yeah, your explanation here makes sense, it just isn't obvious in the code.

**Comment by jakule** [comment]
Maybe related https://github.com/gravitational/teleport/issues/4469

**Comment by zmb3** [comment]
May also be related: #6245

File: lib/srv/reexec.go [hunk 2]
```diff
@@ -189,6 +195,10 @@ func RunCommand() (errw io.Writer, code int, err error) {
 	if contfd == nil {
 		return errorWriter, teleport.RemoteCommandFailure, trace.BadParameter("continue pipe not found")
 	}
+	termiantefd := os.NewFile(TerminateFile, "/proc/self/fd/8")
```
Comment (zmb3) [inline lib/srv/reexec.go:198]: Why `fd/8`?

File: lib/srv/term.go [hunk 3]
```diff
@@ -168,7 +178,13 @@ func (t *terminal) AddParty(delta int) {
...
+func (t *terminal) Run(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
 	var err error
 	// Create the command that will actually execute.
 	t.cmd, err = ConfigureCommand(t.ctx)
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/srv/term.go:190]: I was also confused by it, but `t.ctx` has type `ServerContext` where `ctx` is `context.Context`. I can rename `t.ctx` to indicate that this is "not the context that you're thinking of".

[the remark this answers is not in the document]
Comment (jakule) [reply lib/srv/term.go:190]: We're not using and we shouldn't. This code is executed in the main `teleport` process. If we kill/terminate the child process, we will end up in the same place that we're in right now. Currently, the tree looks like that `teleport -> teleport exec -> bash`. Our current implementation kills `teleport exec`, which causes `bash` to be re-assigned to PID 1. Calling cancel on that process would make no difference as, from what I remember, SIGTERM does nothing to `teleport exec`.
With my changes, when `teleport` want to stop the SSH session, it will inform `teleport exec` to kill `bash`. In this scenario:
1. `bash` and all subprocesses will be stopped (unless started with `nohup`)
2. `teleport exec` has a chance to run some post-process actions https://github.com/gravitational/teleport/blob/master/lib/srv/reexec.go#L411-L416 which are currently skipped

In summary, we're using `(*Process).Kill()` or (*Process).Signal()` to interact with any processes give us more flexibility and we won't end up in that weird scenario again when someone cancels context and kills the wrong process.

File: lib/srv/reexec.go (hunk 2 above)
Comment (jakule) [reply lib/srv/reexec.go]: Because it was free, and Linux didn't complain about it. It should be `5` 🙈 
I changed the code to use the same constants.

**Review by zmb3 — APPROVED** [review]

**Review by LKozlowski — APPROVED** [review]
